File: 0040_combinationSumII/main.cpp
#include <vector>
#include <algorithm>
class Solution {
private:
    void backtrack(std::vector<int>& candidates, std::vector<std::vector<int>>& res,
                std::vector<int> subset, int startIdx, int target) {
        
        if(target < 0)
            return;
        if(target == 0) {
            res.push_back(subset);
            return;
        }

        for(int i = startIdx; i < candidates.size(); ++i) {
            if(i > startIdx && candidates[i] == candidates[i-1])
                continue;
            subset.push_back(candidates[i]);
            backtrack(candidates, res, subset, i + 1, target - candidates[i]);
            subset.pop_back();
        }

    }
public:
    std::vector<std::vector<int>> combinationSum2(std::vector<int>& candidates, int target) {
        std::vector<std::vector<int>> res;
        std::vector<int> subset;
        std::sort(candidates.begin(), candidates.end());
        backtrack(candidates, res, subset, 0, target);
        return res;
    }
};

File: 0040_combinationSumII/main.py
from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()

        def backtrack(startIdx, target, subset):
            if target < 0:
                return
            
            if target == 0:
                res.append(subset[:])
                return
            
            for i in range(startIdx, len(candidates)):
                if i > startIdx and candidates[i] == candidates[i-1]:
                    continue
                subset.append(candidates[i])
                backtrack(i + 1, target - candidates[i], subset)
                subset.pop()

        backtrack(0, target, [])
        return res
    
class Solution1:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()

        def backtrack(startIdx, target, subset):
            if target == 0:
                res.append(subset[:])
                return
            
            if startIdx >= len(candidates) or target < 0:
                return
            
            subset.append(candidates[startIdx])
            backtrack(startIdx + 1, target - candidates[startIdx], subset)
            subset.pop()

            while startIdx + 1 < len(candidates) and candidates[startIdx] == candidates[startIdx + 1]:
                startIdx += 1
            backtrack(startIdx + 1, target, subset)

        backtrack(0, target, [])
        return res    

